### PR TITLE
HPCC-31283 Use PING mechanism to detect mismatched configuration files

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
Compiled

### CODE REVIEW

File: roxie/ccd/ccdmain.cpp [hunk 1]
```diff
@@ -464,6 +464,16 @@ static std::vector<std::pair<unsigned, unsigned>> agentChannels;
 
 void *leakChecker = nullptr;   // Used to deliberately leak an allocation to ensure leak checking is working
 
+unsigned __int64 currentTopologyHash = 0;
+unsigned __int64 originalTopologyHash = 0;
+
+hash64_t getTopologyHash()
+{
+    StringBuffer xml;
+    toXML(topology, xml, 0, XML_SortTags);
```
Comment (ghalliday) [inline roxie/ccd/ccdmain.cpp:473]: Should this somehow exclude lockDali from the hash?  Will that cause any problems?

File: roxie/ccd/ccdstate.cpp [hunk 2]
```diff
@@ -3068,6 +3068,7 @@ class CRoxiePackageSetManager : implements IRoxieQueryPackageManagerSet, impleme
             unknown = true;
             break;
         }
+        currentTopologyHash = getTopologyHash();
```
Comment (ghalliday) [inline roxie/ccd/ccdstate.cpp:3071]: Should this only be done if topology has changed?  Does the topology ever get large enough that the time to generate will be significant.

**Review by ghalliday — COMMENTED** [review]
A few trivial questions.  Also 1st commit message makes it sound like you are adding that feature.
```
Prevent control:lockDali "burning in" ephemeral roxie changes
```
might be better.

File: roxie/ccd/ccdstate.cpp (hunk 2 above)
Comment (richardkchapman) [reply roxie/ccd/ccdstate.cpp:3071]: It's not huge, and this only happens on a control: message. I doubt it is detectable, but I don't mind changing if you want me to.

File: roxie/ccd/ccdmain.cpp (hunk 1 above)
Comment (richardkchapman) [reply roxie/ccd/ccdmain.cpp:473]: The intention was for this one to include lockDali, and the other one to exclude it, so we can see both

File: roxie/ccd/ccdstate.cpp (hunk 2 above)
Comment (ghalliday) [reply roxie/ccd/ccdstate.cpp:3071]: I doubt it is matters.  Leave as it is.

**Review by mckellyln — COMMENTED** [review]
I think this looks ok. but I am not 100% clear on how it works.
Does each ping cycle compute, collect and compare topo hash with source ?

**Comment by richardkchapman** [comment]
> I think this looks ok. but I am not 100% clear on how it works.
> Does each ping cycle compute, collect and compare topo hash with source ?

The hash is calculated only when it changes. It is compared on each cycle.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**